### PR TITLE
Adopt DensityInterface.jl

### PR DESCRIPTION
**Pull request by phipsgabler**
This is in the hope of finally having a shared logdensity function across "probabilistic packages". Distributions.jl is already pretty much on board, Soss.jl hopefully will join later as soon as questions about how to integrate this with properly done measure theory (implicit/explicit base measures) have been resolved.

The consensus was to let the name be `logdensityof` (and `densityof`) to avoid name clashes. The interface is intended for both types that "have a density" and "are a density". 

The other part of DensityInterface.jl is a trait, `hasdensity`, which for `AbstractProbabilisticModel` I have defaulted to `true`. I think likelihood-free inference is out of scope for the kind of models we target -- is that a reasonable assumption?

See https://github.com/JuliaStats/Distributions.jl/pull/1416, https://github.com/JuliaMath/DensityInterface.jl/issues/1 and https://github.com/JuliaMath/DensityInterface.jl/issues/3.

### CODE REVIEW

**Comment by yebai** [comment]
> I think likelihood-free inference is out of scope for the kind of models we target -- is that a reasonable assumption?

That should be fine for now. We don't have any support for likelihood-free inference yet. But even for likelihood-free inference, one often uses an approximate likelihood/loss function. So we can still use the density interface.

**Review by yebai — APPROVED** [review]

File: interface.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ Therefore, the interface consists of:
 - `condition(::Model, ::Trace) -> ConditionedModel`
 - `decondition(::ConditionedModel) -> GenerativeModel`
 - `sample(::Model, ::Sampler = Exact(), [Int])` (from `AbstractMCMC.sample`)
-- `logdensity(::Model, ::Trace)`
+- `DensityInterface.logdensityof(::Model, ::Trace)`
```
Comment (devmotion) [inline interface.md:32]: Maybe to be consistent with the line above:
```suggestion
- `logdensityof(::Model, ::Trace)` (from `DensityInterface.logdensityof`)
```
And maybe add and export `densityof` as well which is defined automatically?

File: interface.md [hunk 2]
```diff
@@ -198,32 +198,34 @@ Not all variants need to be supported – for example, a posterior model might n
...
 Since the different “versions” of how a model is to be understood as generative or conditioned are
 to be expressed in the type or dispatch they support, there should be no need for separate functions
-`logjoint`, `loglikelihood`, etc., which force these semantic distinctions on the implementor; one
-`logdensity` should suffice for all, with the distinction being made by the capabilities of the
-concrete model instance.
+`logjoint`, `loglikelihood`, etc., which force these semantic distinctions on the implementor; we
+therefore adapt the interface of
+[DensityInterface.jl](https://github.com/JuliaMath/DensityInterface.jl).  Its one function
+`logdensityof` should suffice for variants, with the distinction being made by the capabilities of
+the concrete model instance.
```
Comment (devmotion) [inline interface.md:209]: It's actually two functions (or three, if one includes `logfuncdensity` which seems irrelevant here though), maybe mention `densityof` as well?

File: src/AbstractPPL.jl [hunk 3]
```diff
@@ -5,7 +5,7 @@ export VarName, getsym, getindexing, getlens, inspace, subsumes, varname, vsym,
 
 
 # Abstract model functions
-export AbstractProbabilisticProgram, condition, decondition, logdensity
+export AbstractProbabilisticProgram, condition, decondition, logdensityof
```
Comment (devmotion) [inline src/AbstractPPL.jl:8]: Maybe
```suggestion
export AbstractProbabilisticProgram, condition, decondition, logdensityof, densityof
```

File: src/abstractprobprog.jl [hunk 4]
```diff
@@ -8,19 +9,21 @@ Common base type for models expressed as probabilistic programs.
...
 
 `trace` can be of any supported internal trace type, or a fixed probability expression.
 
-`logdensity` should interact with conditioning and deconditioning in the way required by probability
-theory.
+`logdensityof` should interact with conditioning and deconditioning in the way required by
+probability theory.
 """
-function logdensity end
+DensityInterface.logdensityof
```
Comment (devmotion) [inline src/abstractprobprog.jl:26]: I think it would be better to add a docstring only to `logdensityof(::AbstractProbabilisticProgram, ::AbstractModelTrace)`.
```suggestion
function DensityInterface.logdensityof(::AbstractProbabilisticProgram, ::AbstractModelTrace) end
```

**Review by devmotion — COMMENTED** [review]
We should also update the version number :slightly_smiling_face:

File: src/abstractprobprog.jl (hunk 4 above)
Comment (torfjelde) [reply src/abstractprobprog.jl:26]: Should we not also deprecate `logdensity` rather than just dropping it completely?

Comment (devmotion) [reply src/abstractprobprog.jl:26]: I used this form since `function f end` form is the recommended syntax in Julia documentation: https://docs.julialang.org/en/v1/manual/documentation/#Functions-and-Methods-2
This also adds a method to `f`.

It seems one can avoid this (even though it seems the other form is recommended) by specifying
```suggestion
DensityInterface.logdensityof(::AbstractProbabilisticProgram, ::AbstractModelTrace)
```
instead.

File: interface.md [hunk 5]
```diff
@@ -198,32 +200,45 @@ Not all variants need to be supported – for example, a posterior model might n
...
 
-Note that this generalizes `logpdf`, too, since the posterior density will of course in general be
-unnormalized and hence not a probability density.
+ DensityInterface.jl also requires the trait function `hasdensity`, which is set to `true` for the
+`AbstractProbabilisticProgram` type.  Additional functions
 
-The evaluation will usually work with the internal, concrete trace type, like `VarInfo` in Turing.jl:
+```
+DensityInterface.densityof(d, x) = exp(logdensityof(d, x))
+DensityInterface.logdensity(d) = Base.Fix1(logdensity, d)
```
Comment (mschauer) [inline interface.md:218]: Is this correct?

File: interface.md [hunk 6]
```diff
@@ -24,12 +24,14 @@ There are three interrelating aspects that this interface intends to standardize
...
 
-Therefore, the interface consists of:
+Therefore, the interface consists of an `AbstractProbabilisticProgram` supertype, together with
+functions
 
 - `condition(::Model, ::Trace) -> ConditionedModel`
 - `decondition(::ConditionedModel) -> GenerativeModel`
 - `sample(::Model, ::Sampler = Exact(), [Int])` (from `AbstractMCMC.sample`)
-- `logdensity(::Model, ::Trace)`
+- `logdensityof(::Model, ::Trace)` and `density(::Model, ::Trace)` (from
```
Comment (mschauer) [inline interface.md:33]: Is this correct?

File: src/AbstractPPL.jl [hunk 7]
```diff
@@ -5,7 +5,7 @@ export VarName, getsym, getindexing, getlens, inspace, subsumes, varname, vsym,
 
 
 # Abstract model functions
-export AbstractProbabilisticProgram, condition, decondition, logdensity
+export AbstractProbabilisticProgram, condition, decondition, logdensityof, densityof
```
Comment (mschauer) [inline src/AbstractPPL.jl:8]: Is densityof defined somewhere?

Comment (phipsgabler) [reply src/AbstractPPL.jl:8]: Defaulted in DensityInterface, yes. https://github.com/JuliaMath/DensityInterface.jl/blob/ccade11dec6920cbc19852650e565d10254f1285/src/interface.jl#L156

File: interface.md (hunk 6 above)
Comment (mschauer) [reply interface.md:33]: No, it should be `densityof`

File: interface.md (hunk 5 above)
Comment (mschauer) [reply interface.md:218]: No it should be `logdensityof`

File: src/AbstractPPL.jl (hunk 7 above)
Comment (mschauer) [reply src/AbstractPPL.jl:8]: Ok

**Comment by devmotion** [comment]
Yes, this is one of the main reasons for a unified interface. However, it will require some changes in DynamicPPL since we don't use `logpdf` currently but `loglikelihood` due to its implicit IID assumptions and support of multiple observations.

**Comment by mschauer** [comment]
That's why I am asking, in the current form of https://github.com/JuliaStats/Distributions.jl/pull/1416 using `IIDDensity` in Turing will mean that the `logdensityof` interface is only supported for `B<:Distribution` because `IIDDensity` lives in Distributions.jl

File: interface.md (hunk 5 above)
Comment (phipsgabler) [reply interface.md:218]: Ah, right. I'm still getting used to the `of`...

Comment (mschauer) [reply interface.md:218]: I was confused too

**Comment by devmotion** [comment]
Yes, that's something we have to address in the future in some way 🙂 Currently, Distributions is baked in in DynamicPPL and Turing very deeply, so it will take some time and effort to generalize it regardless of `IIDDensity`.

**Comment by phipsgabler** [comment]
TBH, I had no idea we used `loglikelihood`. Is that the use case? Is it used only for observations/conditioned variables?

```julia
julia> @model function foo(x)
           m ~ Normal()
           x ~ Normal(m)
       end
foo (generic function with 2 methods)

julia> foo(1.2)()
1.2

julia> foo([1.2, 0.2])()
2-element Vector{Float64}:
 1.2
 0.2
```

Is this really a good thing? It definitely makes the type checker in my brain complain.

**Comment by devmotion** [comment]
It is a good idea since before we had to maintain all kinds of custom dispatches for different distributions (but of course would miss any others) and always allocated arrays for the individual log densities which then were summed immediately. It is much more sane to let it implement whoever defines a Distribution type.

**Comment by devmotion** [comment]
But I would suggest not polluting this PR with the seemingly unrelated discussion 🙂

**Comment by mschauer** [comment]
"Polluting" perhaps don't use that loaded language.

**Comment by mschauer** [comment]
No problem, I just wanted to signal that the statement touched a nerve with me on the first take.

File: Project.toml [hunk 8]
```diff
@@ -3,13 +3,15 @@ uuid = "7a57a42e-76ec-4ea3-a279-07e840d6d9cf"
 keywords = ["probablistic programming"]
 license = "MIT"
 desc = "Common interfaces for probabilistic programming"
-version = "0.3.1"
+version = "0.4"
```
Comment (devmotion) [inline Project.toml:6]: In principle the PR doesn't have to be breaking, you could just deprecate `logdensity`. But I guess DynamicPPL is the only/main dependency of AbstractPPL currently, so it doesn't really matter.

Comment (phipsgabler) [reply Project.toml:6]: As we are still in beta, and want to promote DensityInterface adoption, I prefer breaking.  We could leave `logdensity` deprecated in DPPL, though, I don't know how long it has been around before.

Comment (devmotion) [reply Project.toml:6]: Sounds good!
